qiskit/__init__.py

```diff
@@ -14,9 +14,36 @@
 
 """Main Qiskit public functionality."""
 
+import importlib.metadata
+import importlib.util
+import os
 import sys
 import warnings
 
+try:
+    importlib.metadata.version("qiskit-terra")
+except importlib.metadata.PackageNotFoundError:
+    # All good!
+    pass
+else:
+    # 'qiskit.tools' is present in all 0.x series of Qiskit and not in Qiskit 1.0+.  If a dev has an
+    # editable install and switches from 0.x branches to 1.0+ branches, they might have an empty
+    # `qiskit/tools` folder in their path, which will appear as a "namespace package" with no valid
+    # location.  We catch that case as "not actually having Qiskit 0.x" as a convenience to devs.
+    _has_tools = getattr(importlib.util.find_spec("qiskit.tools"), "has_location", False)
```

Comment:
This is a clever trick relying on the 1.0 removals to determine if the terra import is breaking anything or not.

Reply:
Thanks!

Will (@wshanks) alluded to a different case we might want to catch / improve the documentation to handle over in https://github.com/Qiskit/documentation/pull/710#discussion_r1472323844 in addition to this.  If he's got time to find the repro steps he mentions I can see if there's anything we might want to do in the import mechanism.

I don't want it to get too complex, though - I'd rather have a couple of (minor) false negatives for devs than slow imports or false positives for all users

Reply:
I wasn't able to reproduce it by installing 0.25->0.45->main, so I don't think we need to worry about it. Either it takes an intricate set of steps to reproduce or I was just confused about something when I saw it.

Reply:
Ok cool, thanks Will.  If you do come across another case, let us know - it's possible for us to update this check without much pain (though I am _really_ nervous about how I've been deliberately breaking `import`!)